spring-integration-ip/src/main/java/org/springframework/integration/ip/config/TcpConnectionFactoryFactoryBean.java

```diff
@@ -128,8 +137,11 @@ public final void setBeanFactory(BeanFactory beanFactory) {
 
 	@Override
 	public Class<?> getObjectType() {
-		return this.connectionFactory != null ? this.connectionFactory.getClass()
-											  : AbstractConnectionFactory.class;
+		return this.connectionFactory != null ? this.connectionFactory.getClass() :
+			this.type == null ? AbstractConnectionFactory.class :
+				isServer() ? AbstractServerConnectionFactory.class :
+					isClient() ? AbstractClientConnectionFactory.class :
+						AbstractConnectionFactory.class;
```

Comment:
It's not so easy because the constructors are different - and this wouldn't prevent a user passing in a subclass which has completely different constructors.

TBH, this BF is really only useful for XML - for JavaConfig you can simply instantiate the desired CF directly - see https://github.com/spring-cloud/spring-cloud-stream-modules/pull/209/commits/3aaa55ec1d23e5af84f9b298b20515bde692e2ab